lib/worker.go

```diff
@@ -672,9 +707,97 @@ func (m *masterClient) SendHeartBeat(ctx context.Context, clock clock.Clock) err
 	return nil
 }
 
+func (m *masterClient) IsMasterSideClosed() bool {
+	return m.masterSideClosed.Load()
+}
+
 // used in unit test only
 func (m *masterClient) getLastMasterAckedPingTime() clock.MonotonicTime {
 	m.mu.RLock()
 	defer m.mu.RUnlock()
 	return m.lastMasterAckedPingTime
 }
+
+const (
+	workerExitWaitForMasterTimeout = time.Second * 15
+)
+
+// workerExitController implements the exit sequence of
+// a worker. This object is thread-safe.
+// TODO move this to a separate file or package.
+type workerExitController struct {
+	workerExitFsm atomic.Int32
+	halfExitTime  atomic.Time
+	errCenter     *errctx.ErrCenter
+	masterClient  *masterClient
+
+	// clock is to facilitate unit testing.
+	clock clock.Clock
+}
+
+func newWorkerExitController(
+	masterClient *masterClient,
+	errCenter *errctx.ErrCenter,
+	clock clock.Clock,
+) *workerExitController {
+	return &workerExitController{
+		workerExitFsm: *atomic.NewInt32(workerNormal),
+		errCenter:     errCenter,
+		masterClient:  masterClient,
+		clock:         clock,
+	}
+}
+
+// PollExit is called in each tick of the worker.
+// Returning an error other than ErrWorkerHalfExit
+// means that the worker is ready to exit.
+func (c *workerExitController) PollExit() error {
+	err := c.errCenter.CheckError()
+	if err == nil {
+		return nil
+	}
+
+	if derror.ErrWorkerSuicide.Equal(err) {
+		// Suicides should result in an immediate exit.
+		c.workerExitFsm.Store(workerExited)
+		return err
+	}
+
+	switch c.workerExitFsm.Load() {
+	case workerNormal:
+		c.workerExitFsm.Store(workerHalfExit)
+		c.halfExitTime.Store(c.clock.Now())
+		return derror.ErrWorkerHalfExit.FastGenByArgs()
+	case workerHalfExit:
+		if c.masterClient.IsMasterSideClosed() {
+			c.workerExitFsm.Store(workerExited)
+			return err
+		}
+		sinceStartExiting := c.clock.Since(c.halfExitTime.Load())
+		if sinceStartExiting > workerExitWaitForMasterTimeout {
```

Comment:
I have a concern for this timeout(15s) here.
We have the guarantee previously that master need recreate a new worker before worker commit suicide.
Do you have other assert to check that guarantee? Or we will need to wait at most WorkerTimeoutDuration+ workerExitWaitForMasterTimeout (15s+15s) if workerExitFsm(Exit) is overrided by  PollExit logic.

Reply:
> We have the guarantee previously that master need recreate a new worker before worker commit suicide.

The master can recreate the worker **after** the worker commits suicide. It is the master's side that needs to wait longer to ensure safety.

Reply:
My mistake for the wrong statement. It's `After`.
We need some assertion to guarantee the `The master can recreate the worker after the worker commits suicide.`, like :https://github.com/hanfei1991/microcosm/blob/c16e9f05bfcbf90ecb6a48be9210b3bfa8bcc5c1/lib/config/timeouts.go#L23

Reply:
Sometimes we may add other timeout in this critical path which violate the guarantee, like the `workerExitWaitForMasterTimeout` here will increase the worker exit timespan.

Reply:
Consider such time sequence：
t0: doPoll() -> PollExit() -> get app error -> enter workerStateNormal branch
t1: runWatchDog ->  workerTimeout -> set workerStateExit
t2: overide and set workerState to workerHalfExit
t3: wait workerExitWaitForMasterTimeout(15s) to exit

Because a time slice in Load() and Store()

Reply:
resolved in the latest commit.